crates/task-impls/src/quorum_proposal_recv.rs

```diff
@@ -0,0 +1,409 @@
+
+                    let state = Arc::new(
+                        <TYPES::ValidatedState as ValidatedState<TYPES>>::from_header(
+                            &proposal.data.block_header,
+                        ),
+                    );
+
+                    consensus.validated_state_map.insert(
+                        view,
+                        View {
+                            view_inner: ViewInner::Leaf {
+                                leaf: leaf.commit(),
+                                state,
+                                delta: None,
+                            },
+                        },
+                    );
+                    consensus.saved_leaves.insert(leaf.commit(), leaf.clone());
+
+                    if let Err(e) = self
+                        .storage
+                        .write()
+                        .await
+                        .update_undecided_state(
+                            consensus.saved_leaves.clone(),
+                            consensus.validated_state_map.clone(),
+                        )
+                        .await
+                    {
+                        warn!("Couldn't store undecided state.  Error: {:?}", e);
+                    }
+
+                    // If we are missing the parent from storage, the safety check will fail.  But we can
+                    // still vote if the liveness check succeeds.
+                    let liveness_check = justify_qc.get_view_number() > consensus.locked_view;
+
+                    let high_qc = consensus.high_qc.clone();
+                    let locked_view = consensus.locked_view;
+
+                    drop(consensus);
+
+                    if liveness_check {
+                        self.current_proposal = Some(proposal.data.clone());
+                        let new_view = proposal.data.view_number + 1;
+
+                        // This is for the case where we form a QC but have not yet seen the previous proposal ourselves
+                        let should_propose = self.quorum_membership.get_leader(new_view)
+                            == self.public_key
+                            && high_qc.view_number
+                                == self.current_proposal.clone().unwrap().view_number;
+                        let qc = high_qc.clone();
+                        if should_propose {
+                            debug!(
+                                "Attempting to publish proposal after voting; now in view: {}",
+                                *new_view
+                            );
+                            let pdd = {
+
+                            }
+                            broadcast_event(
+                                Arc::new(HotShotEvent::ProposeNow(
+                                    qc.view_number + 1,
+
+                                ))
+                            )
+                            // self.publish_proposal_if_able(qc.view_number + 1, &event_stream)
```

Comment:
Looks like the code inside `publish_proposal_if_able` isn't pasted to here yet. We can do that later, but I just wanted to leave this comment in case we forget.